### PR TITLE
fixed: attachment support was missing in CPTextView

### PR DESCRIPTION
**Pull request by daboe01**
Previously, CPTextView was only able to handle text.
This PR introduces support for embedding CPViews into the text.
You can use this via the CPTextStorage method ```-attributedStringWithAttachment:(CPView)someView ```
These are treated as individual characters with their respective rectangular sizes.

### CODE REVIEW

**Comment by daboe01** [comment]
+TextView
+#needs-improvement

**Comment by daboe01** [comment]
-#new

**Comment by jasperdeb** [comment]
Nice work. 

To have a good user experience, I think it required that an image can be pasted into the CPTextView.

A few of the fundamental questions are: 
- Is it necessary to upload an image to the webserver (in the Resources folder) to display the image, in order to draw it to the screen?
For example I think that Gmail does upload a pasted image.
- If an upload to the server is needed, can we avoid to introduce a dependency on for example the Cup framework?

**Comment by daboe01** [comment]
@jasperdeb yes, the gui needs to support the stuff you mentioned. this PR is to update the model and controller to support this in the GUI. the image can obviously also drawn via data url / canvas.

**Comment by jasperdeb** [comment]
Ok those issues are maybe for a next PR. To have some intermediate success we should first indeed make sure we can draw the image.

File: AppKit/CPTextView/CPTypesetter.j [hunk 1]
```diff
@@ -322,6 +323,23 @@ var CPSystemTypesetterFactory,
...
+
+                 if (imageSize.heigth > _lineBase)
+                    _lineBase = imageSize.heigth;
+
+                 isAttachment = YES;
+                 wrapRange = CPMakeRangeCopy(lineRange);
+                 wrapWidth = rangeWidth;
+                 wrapRange._height = _lineHeight;
+                 wrapRange._base = _lineBase;
+                 break;
```
Comment (jasperdeb) [inline AppKit/CPTextView/CPTypesetter.j]: Missing right curly bracket to close close the CPAttachmentCharacter

File: AppKit/CPTextView/CPTypesetter.j [hunk 2]
```diff
@@ -322,6 +323,23 @@ var CPSystemTypesetterFactory,
 
         switch (currentCharCode)    // faster than sending actionForControlCharacterAtIndex: called for each char.
         {
+            case CPAttachmentCharacter:
+            {
+                var attributes = [_textStorage attributesAtIndex:glyphIndex effectiveRange:nil],
+                    imageSize = CGSizeFromString([attributes objectForKey:_CPAttachmentImageSize],
```
Comment (jasperdeb) [inline AppKit/CPTextView/CPTypesetter.j]: Missing right parenthesis to close CGSizeFromString

**Comment by jasperdeb** [comment]
image

A first result working ... (not yet completely functional)

File: AppKit/CPTextView/CPLayoutManager.j [hunk 3]
```diff
@@ -1147,17 +1161,29 @@ var _objectsInRange = function(aList, aRange)
...
+
+            // this is an attachment -> create a run for this
+            if (string === String.fromCharCode(CPAttachmentCharacter))
+            {
+               // <!> fixme
+               // var imageExtractedFromAttachment = ...
+               // var elem = [self createDOMElementWithImage:imageExtractedFromAttachment],
+               //     run = {_range:CPMakeRangeCopy(effectiveRange), color:color, font:font, elem:nil, string:string};
+
+               // _runs.push(run);
```
Comment (jasperdeb) [inline AppKit/CPTextView/CPLayoutManager.j]:
```
var imageExtractedFromAttachment = [[CPImage alloc] initWithContentsOfFile:[attributes objectForKey:_CPAttachmentImageFile] size:[attributes objectForKey:_CPAttachmentImageSize]];
 var elem = [self createDOMElementWithImage:imageExtractedFromAttachment],
 run = {_range:CPMakeRangeCopy(effectiveRange), color:nil, font:nil, elem:elem, string:nil};

 _runs.push(run);
```

I have added this code for the moment

File: AppKit/CPTextView/CPLayoutManager.j [hunk 4]
```diff
@@ -1125,6 +1125,20 @@ var _objectsInRange = function(aList, aRange)
...
 }
+- (id)createDOMElementWithImage:(CPImage)anImage
+{
+#if PLATFORM(DOM)
+    var style,
+        img = document.createElement("img");
+
+    img.oncontextmenu = img.onmousedown = img.onselectstart = _oncontextmenuhandler;
+    // fixme: config the img accoringly
+
```
Comment (jasperdeb) [inline AppKit/CPTextView/CPLayoutManager.j]:
```
var style,
        img = document.createElement("img"),
        span = document.createElement("span");

    style = span.style;
    style.position = "absolute";
    style.visibility = "visible";
    style.padding = "0px";
    style.margin = "0px";
    style.whiteSpace = "pre";
    style.backgroundColor = "transparent";

    img.oncontextmenu = img.onmousedown = img.onselectstart = _oncontextmenuhandler;
    // fixme: config the img accoringly

    img.src = [anImage filename];
    img.height = [anImage size].height;
    img.width = [anImage size].width;

    span.innerHTML = "<img src='"+[anImage filename]+"' height='"+[anImage size].height+"' width='"+[anImage size].width+"'>";

    //span.appendChild(img);

    return span;
```

This is not yet the code that it should be, but at the moment I tried something to make things work ...

**Comment by jasperdeb** [comment]
@daboe01 what do you think of this last commit? I think it is more logical. 

Nevertheless when I make a simple example with text and images, and then add text at the beginning of the CPTextView, the code crashes when the text should be wrapped as the image does not fit on the screen anymore... I do not seem to be able to solve this. Can you take a look?

**Comment by daboe01** [comment]
@jasperdeb thank you! great progress.

**Comment by daboe01** [comment]
@jasperdeb i am reluctant towards c79166a, i.e. if it does not fix your issue.

**Comment by daboe01** [comment]
the problem with c79166a was that we need the advances uncoditionally for the layoutmanager to work properly

**Comment by jasperdeb** [comment]
@daboe01  I think we are in quite good shape for displaying an image in the layoutmanager. 

Do we use this PR for the following steps (adding image support to RFT-parsing for drag and drop and copy paste, upload new images from the client side, ...)? 
What is your idea to continue from this point?

**Comment by daboe01** [comment]
@jasperdeb great!
i think it should be possible to handle intra-app copy/paste and drag/drop with a dedicated pasteboard type that uses e.g. CPCoder.

dealing with the browser-pasteboard is a different story though.
see e.g. here: https://ourcodeworld.com/articles/read/491/how-to-retrieve-images-from-the-clipboard-with-javascript-in-the-browser

**Comment by daboe01** [comment]
@jasperdeb i want to use this stuff for fixing CPTokenField.

**Comment by daboe01** [comment]
@jasperdeb is editing with images completely flawless? or do you still see issues?
if not so, i would flag this PR as ready-to-commit

**Comment by jasperdeb** [comment]
@daboe01  I will do some more testing this evening. 

Personally I think that the intra app copy/paste and drag and drop should be a part of this PR. Not having that will create some loose ends (fatal errors) for users if we already ask this to merge.

**Comment by daboe01** [comment]
@jasperdeb i will look into this.
but keep in mind that attachments are not exposed by default.

**Comment by jasperdeb** [comment]
I have found a few issues (will try to help to solve them later in the week):

- There is a crash when an image is inserted which is bigger then the frame of the CPTextImage 
For example it crashes when trying to insert
```
var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/louvre.jpg" size:CGSizeMake(450, 300)];
    [textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
```
into a CPTextView
`textView = [[CPTextView alloc] initWithFrame:CGRectMake(5, 5, 1500, 190)];`

- If you insert a square image, the caret is not visible when before the image. Maybe we should define the glyph just a little bigger than the image width? In the example below the caret is on the red location, (but behind the image).

image

- If you insert a bigger image then a previous image on the same line, the baseline is not correctly calculated.
image

- If you press 2 times enter directly after an image. The second time it will go much too low with the new line. It still goes the height of the image down, and not the height of the empty text line below the image.
image

**Comment by jasperdeb** [comment]
A deviation for cocoa (texteditor). The caret is the size of the biggest character on the line. In CPTextView te caret changes in height from each glyph on a same line

image

**Comment by daboe01** [comment]
@jasperdeb 
thank you for testing.
i will look into these issues as soon as possible.
 i know the caret height is a deviation from cocoa, but prefer it the way it is right now.

**Comment by daboe01** [comment]
@jasperdeb i fixed a lot of stuff. are there any issues left?

**Comment by jasperdeb** [comment]
@daboe01 Most issues seem to be fixed! Great work! The drag-and drop glitch and the crash when an image is inserted which is wider then the width of the CPTextView seem to remain.

**Comment by jasperdeb** [comment]
This fix is not the default behaviour of for example Text Editor on MacOS, but for the moment we need this not to crash the app. 

I think we also need to add some interface to adjust the image size.

**Comment by daboe01** [comment]
@jasperdeb thank you for testing.
2333425 does not look right to me.
i think that the image should be clipped an not be resized.
the fix should be in the typesetter/layoutmanager,

**Comment by jasperdeb** [comment]
Feel free to suggest an idea to fix this issue in typesetter/layoutmanager and clipping the image.

**Comment by daboe01** [comment]
@jasperdeb i will look into the remaining issues asap.

**Comment by daboe01** [comment]
@jasperdeb do you still see issues after my latest patches?

**Comment by jasperdeb** [comment]
The clipped image does work as expected. The drag placeholder issue is not yet solved for me.

**Comment by daboe01** [comment]
@jasperdeb can you please post a screenshot that exposes the drag placeholder issue?

**Comment by jasperdeb** [comment]
image

**Comment by jasperdeb** [comment]
I have selected tesxt from "sdf..." threfore the image should not be in the placeholder, and the text after a space is on the next line, which should not be like that.

**Comment by daboe01** [comment]
@jasperdeb thank you! that helps. i will fix df7a000 asap.

**Comment by daboe01** [comment]
@jasperdeb fixing this was harder than first thought :-)
do you still see issues?

**Comment by jasperdeb** [comment]
When I select some glyps ( so they are marked with a blue background), I am nog able to select these glyps to start to drag. 
The start of the dragging initiates a new glyph selection.

**Comment by jasperdeb** [comment]
image

Also the text after an image is on the next line, which also should be like that I think.

**Comment by daboe01** [comment]
@jasperdeb thank you for spotting this with the character next to the image. this issue should be fixed now.
i did not manage so far to reproduce your character-on-next-line issue.

**Comment by daboe01** [comment]
@jasperdeb can you please check whether 178df37 fixes your wrapping issue?

**Comment by jasperdeb** [comment]
The placeholders have no issues anymore after the fixes, fantastic work! 
... except of something I just saw. In Text Editor on Max OS X it is not possible to select a dragging location inside the selected text. 

The caret should not be shown at the location below.
image

**Comment by jasperdeb** [comment]
I found something additional. When you add the following code:

```
[textView insertText:"123"];
[textView insertText:" 456 "];
[textView insertText:" 789 "];
var centeredParagraph=[CPParagraphStyle new];
    [centeredParagraph setAlignment: CPCenterTextAlignment];
    [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
    attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
     forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
```
you see the following (as expected). The whole paragraph is centered.
image

But if you do the same thing with an image.

```
[textView insertText:"123"];
var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/louvre.jpg" size:CGSizeMake(450, 300)];
[textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
[textView insertText:" 456 "];
var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/Eiffel tower.jpg" size:CGSizeMake(180, 120)];
[textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
[textView insertText:" 789 "];

var centeredParagraph=[CPParagraphStyle new];
    [centeredParagraph setAlignment: CPCenterTextAlignment];
    [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
    attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
     forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
```
The alignment of the paragraph is not correct.

image

**Comment by jasperdeb** [comment]
The build in copy paste fix seems to work (as long you use the right click copy-past and not Command-C and Command -V). The package is starting to look quite good!

**Comment by jasperdeb** [comment]
I have noticed during testing that images are clipped and not wrapped at the end of a line. I think the clipping is only the good behaviour when an image is too large for the CPTextView frame. But when an image fits into the screen, it should be wrapped. (also the behaviour in Text Editor on MacOS.

image

**Comment by jasperdeb** [comment]
Yet another thing to do, is adding a border around the image when hovering over it. Also a button/image in the top right corner should appear as in Text Editor. When clicked on this button this could be a good place to put for example a CPPopover to change the image size.

image

Sorry to make the to do list ever longer :-)

**Comment by daboe01** [comment]
@jasperdeb text attachments must not be images all the time. i want to use them for the tokens in a redone CPTokenField. we should not draw the border and button at the lowest level.

**Comment by daboe01** [comment]
i do not think that deliberately inserting multiple parapgraph styles in the same line makes sense or should be supported.
do you have an idea how macos handles this?

> I found something additional. When you add the following code:
> 
> ```
> [textView insertText:"123"];
> [textView insertText:" 456 "];
> [textView insertText:" 789 "];
> var centeredParagraph=[CPParagraphStyle new];
>     [centeredParagraph setAlignment: CPCenterTextAlignment];
>     [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
>     attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
>      forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
> ```
> 
> you see the following (as expected). The whole paragraph is centered.
> image
> 
> But if you do the same thing with an image.
> 
> ```
> [textView insertText:"123"];
> var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/louvre.jpg" size:CGSizeMake(450, 300)];
> [textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
> [textView insertText:" 456 "];
> var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/Eiffel tower.jpg" size:CGSizeMake(180, 120)];
> [textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
> [textView insertText:" 789 "];
> 
> var centeredParagraph=[CPParagraphStyle new];
>     [centeredParagraph setAlignment: CPCenterTextAlignment];
>     [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
>     attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
>      forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
> ```
> 
> The alignment of the paragraph is not correct.
> 
> image

**Comment by daboe01** [comment]
my next take will be to make all sorts of views work as attachments.

**Comment by jasperdeb** [comment]
> i do not think that deliberately inserting multiple parapgraph styles in the same line makes sense or should be supported.
> do you have an idea how macos handles this?
> 
> > I found something additional. When you add the following code:
> > ```
> > [textView insertText:"123"];
> > [textView insertText:" 456 "];
> > [textView insertText:" 789 "];
> > var centeredParagraph=[CPParagraphStyle new];
> >     [centeredParagraph setAlignment: CPCenterTextAlignment];
> >     [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
> >     attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
> >      forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
> > ```
> > 
> > 
> > you see the following (as expected). The whole paragraph is centered.
> > image
> > But if you do the same thing with an image.
> > ```
> > [textView insertText:"123"];
> > var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/louvre.jpg" size:CGSizeMake(450, 300)];
> > [textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
> > [textView insertText:" 456 "];
> > var tempImag = [[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/Eiffel tower.jpg" size:CGSizeMake(180, 120)];
> > [textView insertText:[CPTextStorage attributedStringWithImageAttachment:tempImag]];
> > [textView insertText:" 789 "];
> > 
> > var centeredParagraph=[CPParagraphStyle new];
> >     [centeredParagraph setAlignment: CPCenterTextAlignment];
> >     [textView insertText:[[CPAttributedString alloc] initWithString:@"Fusce\n"
> >     attributes:[CPDictionary dictionaryWithObjects:[centeredParagraph, [CPFont boldFontWithName:"Arial" size:18], [CPColor redColor]]
> >      forKeys:[CPParagraphStyleAttributeName, CPFontAttributeName, CPForegroundColorAttributeName]]]];
> > ```
> > 
> > 
> > The alignment of the paragraph is not correct.
> > image

I must say that I am not sure how MacOS handles this. But to implement the - (void)setAlignment:(NSTextAlignment)alignment range:(NSRange)range; method in CPTextView this is on my list to investigate.

But in this specific case I have not specified an alignment as I just insert text. Therefore the insertion of text with another alignment should convert the alignment to the last inserted alignment introduced in the paragraph. This is also the behaviour of CPTextView before the support of images.

**Comment by jasperdeb** [comment]
> @jasperdeb text attachments must not be images all the time. i want to use them for the tokens in a redone CPTokenField. we should not draw the border and button at the lowest level.

If the implementation makes more sense on another level, this is ofcourse better. If I understand well the focus is to finish this PR with images, and then generalise this from images to views more in general?

**Comment by jasperdeb** [comment]
When an image is wrapped from a line, but the image is after the wrapped area, the non wrapped part of the line has not the correct line height.

image

**Comment by jasperdeb** [comment]
There seems to be something wrong with the start of the line heights. The leading of the line above is not correctly taken into account for the line below.

image

The big letter y is not completely visible in the first line.

**Comment by jasperdeb** [comment]
The selection of text, the dragging placeholder and the dragging of text seems to work as expected.

**Comment by daboe01** [comment]
my plan is to rework this PR to support something like: attributedStringWithAttachment:(CPView)anAttachment

on this basis we could write a CPImageAttachmentView that handles the resizing button and the like


> > @jasperdeb text attachments must not be images all the time. i want to use them for the tokens in a redone CPTokenField. we should not draw the border and button at the lowest level.
> 
> If the implementation makes more sense on another level, this is ofcourse better. If I understand well the focus is to finish this PR with images, and then generalise this from images to views more in general?

**Comment by daboe01** [comment]
i do not see this clearly on your screenshot. do you have a reduction in code?

> When an image is wrapped from a line, but the image is after the wrapped area, the non wrapped part of the line has not the correct line height.
> 
> image

**Comment by jasperdeb** [comment]
This is simple a line with an image in. And then at the beginning of the line I start to typ some text until the image wraps. If you do not see this, I will post a reduction.

**Comment by daboe01** [comment]
when i "type over" an image the line height looks right to me.
Bildschirmfoto 2019-12-02 um 20 09 23

**Comment by jasperdeb** [comment]
try only text in the first line, and the wrapped image on the second line.

Situation just before wrapping.
image

Situation just after wrapping is applied on the image.
image

**Comment by daboe01** [comment]
@jasperdeb thank you. should be fixed by 97ff2e3.

**Comment by jasperdeb** [comment]
@daboe01 fix 97ff2e3 does the job.

**Comment by daboe01** [comment]
hi gang,
generic view support seems to be working quite well.
Bildschirmfoto 2019-12-04 um 19 15 49
do you see any issues?
if not so, i would like to mark this as ready for commit.

**Comment by jasperdeb** [comment]
@daboe01 wow this look good stuff!

I have added a CPTableView

```
var scrollView1= [[CPScrollView alloc] initWithFrame:CGRectMake(10, 10, 1400, 550)];
    var contentView1 = [scrollView1 contentView];
    
    var textView = [[CPTextView alloc] initWithFrame:CGRectMake(5, 5, 1400, 550)];
    [textView setHorizontallyResizable:NO];
    [textView setDelegate:self];

    [textView setRichText:YES];

    [textView insertText:"This is text to test"];

    var tempImageView = [[CPImageView alloc] initWithFrame:CGRectMake(0, 0, 450, 300)];
    [tempImageView setImage:[[CPImage alloc] initWithContentsOfFile:@"CPTextView_images/louvre.jpg" size:CGSizeMake(450, 300)]]

    [textView insertText:[CPTextStorage attributedStringWithAttachment:tempImageView]];

    [textView insertText:" 456 "];

    // ---- Begin add scrollview
    var scrollView2 = [[CPScrollView alloc] initWithFrame:CGRectMake(0, 0, 300,  200)];
    var contentView2 = [scrollView2 contentView];

    [scrollView2 setBorderType:CPLineBorder];

    [contentView2 setBackgroundColor:[CPColor colorWithRed:212.0 / 255.0 green:221.0 / 255.0 blue:230.0 / 255.0 alpha:1.0]];
    tableView = [[CPTableView alloc] initWithFrame:CGRectMakeZero()];

    theColumn = [[CPTableColumn alloc] initWithIdentifier:@"naam"];
    [[theColumn headerView] setStringValue:@"Test:"];
    [theColumn setWidth:250];
    [tableView addTableColumn:theColumn];

    [textView insertText:[CPTextStorage attributedStringWithAttachment:scrollView2]];

    // ---- End add scrollview
  
    // Add tableView and scrollview to the contentview
    [scrollViewContract setDocumentView:tableViewSC];

    [textView insertText:" 789 qsfqsfd"];

    [scrollViewOPM2 setDocumentView:textView];
```
But I see again the bug where the placeholder of dragging is not correct.

image

**Comment by jasperdeb** [comment]
> There seems to be something wrong with the start of the line heights. The leading of the line above is not correctly taken into account for the line below.
> 
> image
> 
> The big letter y is not completely visible in the first line.

This is not yet solved, but seems to be a more general problem that should be solved in a seperate PR.

**Comment by jasperdeb** [comment]
But I clearly see a new version of the CPTokenField in the close future :-)

**Comment by jasperdeb** [comment]
If you add a CPTableView and hover over the tableView, you don't see the cursor as expected.

image

image

**Comment by daboe01** [comment]
@jasperdeb thank you for testing. i will look into the issues asap.

**Comment by daboe01** [comment]
> But I clearly see a new version of the CPTokenField in the close future :-)

i am also thinking about using this as a replacement for CPCollectionView. CPCollectionView currently has the issue of placing all items into the DOM, even if not exposed.

**Comment by daboe01** [comment]
> There seems to be something wrong with the start of the line heights. The leading of the line above is not correctly taken into account for the line below.
> This is not yet solved, but seems to be a more general problem that should be solved in a seperate PR.

yes, i think that this bug was in before this PR

**Comment by daboe01** [comment]
> If you add a CPTableView and hover over the tableView, you don't see the cursor as expected.

i will have to dig into the complicated field of tracking areas. a baby of @didierkorthoudt .

**Comment by daboe01** [comment]
> But I see again the bug where the placeholder of dragging is not correct.

thank you for spotting. this has to be fixed before we can merge this in.

**Comment by daboe01** [comment]
> > But I see again the bug where the placeholder of dragging is not correct.
> 
> thank you for spotting. this has to be fixed before we can merge this in.

i cannot reproduce. am i missing something?
Bildschirmfoto 2019-12-05 um 22 31 14

**Comment by jasperdeb** [comment]
> > > But I see again the bug where the placeholder of dragging is not correct.
> > 
> > 
> > thank you for spotting. this has to be fixed before we can merge this in.
> 
> i cannot reproduce. am i missing something?
> Bildschirmfoto 2019-12-05 um 22 31 14

I do not yet see a patern. Sometimes it happens, and sometimes it doesn't happen.
image
image

**Comment by jasperdeb** [comment]
But if it happens, it occurs when a part of the last string after the last space is in the selection

**Comment by jasperdeb** [comment]
> > There seems to be something wrong with the start of the line heights. The leading of the line above is not correctly taken into account for the line below.
> > This is not yet solved, but seems to be a more general problem that should be solved in a seperate PR.
> 
> yes, i think that this bug was in before this PR

This is something I will try to look into during the winter holidays at the end of December...

**Comment by daboe01** [comment]
@jasperdeb can you please check, whether 8371123 fixes the drag placeholder issue?

**Comment by jasperdeb** [comment]
@daboe01 At first glance it seems to be fixed!

**Comment by daboe01** [comment]
@jasperdeb thank you for testing

**Comment by daboe01** [comment]
-#needs-improvement
+#ready-to-commit

**Comment by jasperdeb** [comment]
I have taken the 2 classes from the Scrapbook #1 - Views and Layers example on the Cappuccino website, and added this to a CPTextView as

```
[textView insertText:"This is text to test"];

    var scrollView = [[CPScrollView alloc] initWithFrame:CGRectMake(0, 0, 200,  200)];
    var contentView = [scrollView contentView];

    var pageView = [[PageView alloc] initWithFrame:CGRectMake(0, 0, 200.0, 200.0)];
    [pageView setAutoresizingMask:CPViewMinXMargin | CPViewMaxXMargin | CPViewMinYMargin | CPViewMaxYMargin];
    [contentView addSubview:pageView];

    [textView insertText:[CPTextStorage attributedStringWithAttachment:contentView]];

    [textView insertText:" 456 "];
    [textView insertText:" 789 qsfqsfd"];
```

But for some reason I do not see the image drawed in the CPTextView.

image

When I add the view to the main window, the image does appear.

image

@daboe01 Any idea why the image is not appearing?

[Source codes.zip](https://github.com/cappuccino/cappuccino/files/3952632/Source.codes.zip)

**Comment by daboe01** [comment]
@jasperdeb CALayers are currently not supported :-(

**Comment by jasperdeb** [comment]
@daboe01 Ok. Have not yet done something in CALayer, fear I am not able to help in this area ...

**Comment by daboe01** [comment]
@jasperdeb i am not an expert on CALayer myself, but i believe this should be doable. do you have a specific need for this?

**Comment by jasperdeb** [comment]
I am making a view with CALayer to make a crop an image. This is a feature I really like in ms word and outlook, and something I need for a feature I have in mind. I made something which is working started from the scrapbook example on the cappuccino website, which is based on CALayer.

File: AppKit/CPTextView/CPTextStorage.j [hunk 5]
```diff
@@ -257,6 +262,15 @@ var CPTextStorageDelegate_textStorageWillProcessEditing_ = 1 << 1,
     return [super attributedSubstringFromRange:aRange];
 }
 
++ (id)attributedStringWithAttachment:(CPView)someView
+{
+    var  result = [[self alloc] initWithString:_CPAttachmentCharacterAsString];
```
Comment (mrcarlberg) [inline AppKit/CPTextView/CPTextStorage.j]: Small formatting issues! Two spaces after `var`

**Comment by mrcarlberg** [comment]
I have looked at this and I have no issues except the above small formatting issue. Do you guys have anything more before I merge this into the master?

**Comment by jasperdeb** [comment]
I think the CALayer support could be added later. 
Apart from that I don't see any issue with merging this PR.

**Comment by daboe01** [comment]
@mrcarlberg @jasperdeb i advocate to merge this in right now. we will address the missing features later in separate PRs

**Comment by mrcarlberg** [comment]
I have done a quick test of this and I find no problems. It is also isolated to CPTextView so it will not affect other things.

**Comment by mrcarlberg** [comment]
Merged! Thanks!

**Pull request merged**